src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,253 @@
+                                if (referenceEqualsMethod == null || !referenceEqualsMethod.Equals(semanticModel.GetSymbolInfo(receiver, cancellationToken).Symbol))
+                                {
+                                    continue;
+                                }
+
+                                left = arguments[0].Expression;
+                                right = arguments[1].Expression;
+                                break;
+
+                            default:
+                                continue;
+                        }
+
+                        if (!AreOperandsApplicable(left, right, out parameter)
+                            && !AreOperandsApplicable(right, left, out parameter))
+                        {
+                            continue;
+                        }
+
+                        var throwStatement = ifStatement.Statement switch
+                        {
+                            ThrowStatementSyntax @throw => @throw,
+                            BlockSyntax { Statements: { Count: 1 } statements } => statements[0] as ThrowStatementSyntax,
+                            _ => null
+                        };
+
+                        if (throwStatement is null
+                            || throwStatement.Expression is not ObjectCreationExpressionSyntax thrownInIf
+                            || !IsConstructorApplicable(thrownInIf, parameter))
+                        {
+                            continue;
+                        }
+
+                        break;
+
+                    // this.field = param ?? throw new ArgumentNullException(nameof(param));
+                    case ExpressionStatementSyntax
+                    {
+                        Expression: AssignmentExpressionSyntax
+                        {
+                            Right: BinaryExpressionSyntax
+                            {
+                                OperatorToken.RawKind: (int)SyntaxKind.QuestionQuestionToken,
+                                Left: ExpressionSyntax maybeParameter,
+                                Right: ThrowExpressionSyntax { Expression: ObjectCreationExpressionSyntax thrownInNullCoalescing }
+                            }
+                        }
+                    }:
+                        if (!IsParameter(maybeParameter, out parameter) || !IsConstructorApplicable(thrownInNullCoalescing, parameter))
+                        {
+                            continue;
+                        }
+
+                        break;
+
+                    default:
+                        continue;
+                }
+
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    Descriptor,
+                    statement.GetLocation(),
+                    option.Notification.Severity,
+                    additionalLocations: new[] { parameter.Locations[0] },
```

Comment:
for paranoia's sake.  prior to reporting the diag.  ensure:

1. the param has a location.
2. the location's *tree* is the same as our current tree

If so, then you can report this.  

this is likely not necessary, but possibly good in case some error condition makes the compiler collapse symbols in different files.

Reply:
@RikkiGibson Before taking the suggestion, can you add a test similar to this?

```csharp
class C
{
    private int _p;

    public int P
    {
        get => _p;
        set
        {
            local();

            void local()
            {
                _p = value;
            }
        }
    }
}
```

Reply:
I'm adding the suggested test. It passes without issuing any diagnostics, as it should. I do think it's still fine to add the additional validation that Cyrus suggested here.

Reply:
Yup the extra validation doesn't harm and is more future-proof.

Reply:
> 2\. the location's _tree_ is the same as our current tree

I'm assuming this means that the statement should be in the same syntax tree as the parameter.

Reply:
Yup